tests/test_reply_function_sync.js

```diff
@@ -48,152 +33,199 @@ test('get with reply callback returning object', async t => {
+    ({ statusCode, body }) => {
+      t.equal(statusCode, 404)
+      return true
+    }
+  )
   scope.done()
 })
 
-test("when request's content-type is json: reply callback's requestBody should automatically parse to JSON", async t => {
-  const requestBodyFixture = {
-    id: 1,
-    name: 'bob',
-  }
+test('when content-type is json, reply function receives parsed body', async t => {
+  t.plan(3)
+  const exampleRequestBody = JSON.stringify({ id: 1, name: 'bob' })
 
-  const scope = nock('http://service')
-    .post('/endpoint')
+  const scope = nock('http://example.test')
+    .post('/')
     .reply(200, (uri, requestBody) => {
-      t.deepEqual(requestBody, requestBodyFixture)
-      return 'overwrite'
+      t.type(requestBody, 'object')
+      t.deepEqual(requestBody, JSON.parse(exampleRequestBody))
     })
 
-  const { body } = await got.post('http://service/endpoint', {
+  const { statusCode } = await got('http://example.test/', {
     headers: { 'Content-Type': 'application/json' },
-    body: JSON.stringify(requestBodyFixture),
+    body: exampleRequestBody,
   })
-
-  t.equal(body, 'overwrite')
+  t.is(statusCode, 200)
   scope.done()
 })
 
-test("when request has no content-type header: reply callback's requestBody should not automatically parse to JSON", async t => {
-  const requestBodyFixture = {
-    id: 1,
-    name: 'bob',
-  }
+test('without content-type header, body sent to reply function is not parsed', async t => {
+  t.plan(3)
+  const exampleRequestBody = JSON.stringify({ id: 1, name: 'bob' })
 
-  const scope = nock('http://service')
-    .post('/endpoint')
+  const scope = nock('http://example.test')
+    .post('/')
     .reply(200, (uri, requestBody) => {
-      t.deepEqual(requestBody, JSON.stringify(requestBodyFixture))
-      return 'overwrite'
+      t.type(requestBody, 'string')
+      t.equal(requestBody, exampleRequestBody)
     })
 
-  const { body } = await got.post('http://service/endpoint', {
-    body: JSON.stringify(requestBodyFixture),
+  const { statusCode } = await got.post('http://example.test/', {
+    body: exampleRequestBody,
   })
+  t.is(statusCode, 200)
+  scope.done()
+})
+
+// This signature is supported today, however it seems unnecessary. This is
+// just as easily accomplished with a function returning an array:
+// `.reply(() => [200, 'ABC', { 'X-My-Headers': 'My custom header value' }])`
+test('reply with status code, function returning string body, and header object', async t => {
+  const scope = nock('http://example.com')
+    .get('/')
+    .reply(200, () => 'ABC', { 'X-My-Headers': 'My custom header value' })
+
+  const { headers } = await got('http://example.com/')
+
+  t.equivalent(headers, { 'x-my-headers': 'My custom header value' })
 
-  t.equal(body, 'overwrite')
   scope.done()
 })
 
-test('reply can take a callback', async t => {
+test(
+  'reply function returning array with status code',
+  // Seems likely a bug related to https://github.com/nock/nock/issues/1222.
```

Comment:
Is this not the desired behavior? https://github.com/nock/nock/issues/1170

Reply:
Yes, I think it is the desired behavior. However this test doesn't pass. The status code is 202 as expected, but the body is `'[202]'`, while `''` is expected.